File: frontend/src/api/versionComponents.ts
import { get } from './client'

export interface VersionFunction {
  id: string
  name: string
  description: string
  return_type: string
  input_schema: Array<{ name: string; type: string; version_attribute_id: string; required: boolean }>
  version_entity_id: string
  callable_name: string
}

export interface VersionRule {
  id: string
  name: string
  description: string
  condition_expr: string
  priority: string
  input_params: Array<{ name: string; type: string; version_attribute_id: string; required: boolean }>
  version_entity_id: string
}

export interface VersionAction {
  id: string
  name: string
  description: string
  category: string
  action_type: string
  parameters_json: Array<{ name: string; type: string; version_attribute_id: string; required: boolean }>
  version_entity_id: string
}

export interface ImpactAnalysis {
  published_versions: string[]
  referencing_workflows: Array<{ id: string; name: string; version: string }>
  referencing_skills: Array<{ id: string; name: string; version: string }>
  safe_to_delete: boolean
  message: string
}

export function fetchVersionFunctions(versionId: string) {
  return get<VersionFunction[]>(`/ontology-publish/versions/${versionId}/functions`)
}

export function fetchVersionRules(versionId: string) {
  return get<VersionRule[]>(`/ontology-publish/versions/${versionId}/rules`)
}

export function fetchVersionActions(versionId: string) {
  return get<VersionAction[]>(`/ontology-publish/versions/${versionId}/actions`)
}

export function fetchImpactAnalysis(type: 'functions' | 'rules' | 'actions', id: string) {
  return get<ImpactAnalysis>(`/impact-analysis/${type}/${id}`)
}

File: backend/app/models/version_components.py
from datetime import datetime

from sqlalchemy import String, Text, DateTime, ForeignKey, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.utils.identifiers import gen_uuid


class OntologyVersionFunction(Base):
    __tablename__ = "ontology_version_functions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=gen_uuid)
    version_id: Mapped[str] = mapped_column(ForeignKey("ontology_versions.id", ondelete="CASCADE"))
    source_function_id: Mapped[str] = mapped_column(String(36), nullable=False)
    version_entity_id: Mapped[str] = mapped_column(
        ForeignKey("ontology_version_entities.id", ondelete="CASCADE")
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    return_type: Mapped[str] = mapped_column(String(50), default="string")
    input_schema: Mapped[dict | None] = mapped_column(JSON)
    logic_type: Mapped[str] = mapped_column(String(30), default="expression")
    logic_body: Mapped[str] = mapped_column(Text, default="")
    callable_name: Mapped[str] = mapped_column(String(100), default="")
    tags: Mapped[dict | None] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    version: Mapped["OntologyVersion"] = relationship(back_populates="functions")


class OntologyVersionRule(Base):
    __tablename__ = "ontology_version_rules"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=gen_uuid)
    version_id: Mapped[str] = mapped_column(ForeignKey("ontology_versions.id", ondelete="CASCADE"))
    source_rule_id: Mapped[str] = mapped_column(String(36), nullable=False)
    version_entity_id: Mapped[str] = mapped_column(
        ForeignKey("ontology_version_entities.id", ondelete="CASCADE")
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    condition_expr: Mapped[str] = mapped_column(Text, default="")
    conditions_json: Mapped[dict | None] = mapped_column(JSON)
    priority: Mapped[str] = mapped_column(String(10), default="medium")
    input_params: Mapped[dict | None] = mapped_column(JSON)
    output_schema: Mapped[dict | None] = mapped_column(JSON)
    tags: Mapped[dict | None] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    version: Mapped["OntologyVersion"] = relationship(back_populates="rules")


class OntologyVersionAction(Base):
    __tablename__ = "ontology_version_actions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=gen_uuid)
    version_id: Mapped[str] = mapped_column(ForeignKey("ontology_versions.id", ondelete="CASCADE"))
    source_action_id: Mapped[str] = mapped_column(String(36), nullable=False)
    version_entity_id: Mapped[str | None] = mapped_column(
        ForeignKey("ontology_version_entities.id", ondelete="CASCADE"), nullable=True
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    category: Mapped[str] = mapped_column(String(20), default="domain")
    action_type: Mapped[str] = mapped_column(String(30), nullable=False)
    type_config: Mapped[dict | None] = mapped_column(JSON)
    description: Mapped[str | None] = mapped_column(Text)
    parameters_json: Mapped[dict | None] = mapped_column(JSON)
    output_schema: Mapped[dict | None] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    version: Mapped["OntologyVersion"] = relationship(back_populates="actions")


# Avoid circular import — OntologyVersion is defined in version.py
from app.models.version import OntologyVersion  # noqa: E402
